tokio-net/src/driver/sharded_slab/page/mod.rs

```diff
@@ -0,0 +1,198 @@
+use super::Pack;
+use crate::sync::atomic::{spin_loop_hint, AtomicUsize, Ordering};
+
+pub(super) mod slot;
+use self::slot::Slot;
+use std::fmt;
+
+/// A page address encodes the location of a slot within a shard (the page
+/// number and offset within that page) as a single linear value.
+#[repr(transparent)]
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub(super) struct Addr {
+    addr: usize,
+}
+
+pub(super) const INITIAL_SIZE: usize = 32;
+
+const ADDR_INDEX_SHIFT: usize = INITIAL_SIZE.trailing_zeros() as usize + 1;
+
+impl Addr {
+    const NULL: usize = Self::BITS + 1;
+
+    pub(super) fn index(self) -> usize {
+        // Since every page is twice as large as the previous page, and all page sizes
+        // are powers of two, we can determine the page index that contains a given
+        // address by shifting the address down by the smallest page size and
+        // looking at how many twos places necessary to represent that number,
+        // telling us what power of two page size it fits inside of. We can
+        // determine the number of twos places by counting the number of leading
+        // zeros (unused twos places) in the number's binary representation, and
+        // subtracting that count from the total number of bits in a word.
+        super::WIDTH - ((self.addr + INITIAL_SIZE) >> ADDR_INDEX_SHIFT).leading_zeros() as usize
+    }
+
+    pub(super) fn offset(self) -> usize {
+        self.addr
+    }
+}
+
+impl Pack for Addr {
+    const LEN: usize = super::MAX_PAGES + ADDR_INDEX_SHIFT;
+
+    type Prev = ();
+
+    fn as_usize(&self) -> usize {
+        self.addr
+    }
+
+    fn from_usize(addr: usize) -> Self {
+        debug_assert!(addr <= Self::BITS);
+        Self { addr }
+    }
+}
+
+pub(super) type Iter<'a, T> =
+    std::iter::FilterMap<std::slice::Iter<'a, Slot<T>>, fn(&'a Slot<T>) -> Option<&'a T>>;
+
+pub(super) struct Page<T> {
+    prev_sz: usize,
+    remote_head: AtomicUsize,
+    local_head: usize,
+    slab: Box<[Slot<T>]>,
+}
+
+impl<T> Page<T> {
+    pub(super) fn new(size: usize, prev_sz: usize) -> Self {
+        let mut slab = Vec::with_capacity(size);
+        slab.extend((1..size).map(Slot::new));
```

Comment:
I would make this lazy on `insert` if you can. Otherwise, it will be a pretty expensive upfront cost.

Reply:
especially since all the slots may not be used.